### PR TITLE
GF Signup: Dropdown Interval:  Add dropdown to plan interval toggle behind feature flag

### PR DESCRIPTION
**Pull request by ddc22**
- Related Automattic/growth-foundations#296
## Proposed Changes

Adds drop down to the plan interval toggle instead of the radio multi select buttons

image

## Not implemented in this PR

- Sticky behaviour in mobile
- Wiring up the experiment
- Maintain dropdown variant on the `/plans` page
- The clickable area around the options is small and need to be fixed (Might have to dive in to the select component)
- The select does not select an option on enter

## Testing Instructions

* Go to `/start?flags=onboarding%2Finterval-dropdown`
* Make sure the dropdown with term interval appears
* Click on a different interval than yearly and make sure the selected plan stays selected
* Select a monthly plan, checkout and make sure the multi year plan selecter discounts matches the discount shown on the plans page

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

**Comment by isatuncman-auto** [comment]
I tested the user experience and compared the price discounts. Everything seems to be working properly.

Screenshot 2023-12-08 at 2 47 41 PM

**Comment by vinimotaa** [comment]
@jdc91 thanks for sharing it. I wasn't able to reproduce the implementation using `/start?flags=onboarding%2Finterval-dropdown`

From the picture, I noticed that "up to" is written together. Could you add a space in between "up to"?

File: client/my-sites/plans-grid/components/plan-type-selector/hooks/use-interval-options.tsx [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+				...additionalPathProps,
+			} ),
+			discountText: getDiscountText( termViseDiscounts[ key as UrlFriendlyTermType ], translate ),
+		};
+	} );
+
+	return optionList;
+}
+
+// if ( termViseMaxDiscount[ key ] ) {
```
Comment (jeyip) [inline client/my-sites/plans-grid/components/plan-type-selector/hooks/use-interval-options.tsx]: We can probably can delete this 🙂

File: client/my-sites/plans-grid/components/plan-type-selector/components/interval-type-dropdown.tsx [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+		name: (
+			<AddOnOption href={ option.url }>
+				<span className="name"> { option.name } </span>
+				<span className="discount"> { option.discountText } </span>
+			</AddOnOption>
+		),
+	} ) );
+	return (
+		<StyledCustomSelectControl
+			label=""
```
Comment (jeyip) [inline client/my-sites/plans-grid/components/plan-type-selector/components/interval-type-dropdown.tsx:53]: Just verifying that we intentionally want no label?

Comment (ddc22) [reply client/my-sites/plans-grid/components/plan-type-selector/components/interval-type-dropdown.tsx:53]: Yep No label! (if added shows up on the UI which we do not want)

File: client/my-sites/plans-grid/components/plan-type-selector/hooks/use-term-vise-max-dicounts.ts [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+	UrlFriendlyTermType,
+	getPlanMultipleTermsVariantSlugs,
+	getPlanSlugForTermVariant,
+	isMonthly,
+	isWpComPlan,
+	isWpcomEnterpriseGridPlan,
+} from '@automattic/calypso-products';
+import { UsePricingMetaForGridPlans } from 'calypso/my-sites/plans-grid/hooks/npm-ready/data-store/use-grid-plans';
+
+export default function useTermViseMaxDiscounts(
```
Comment (jeyip) [inline client/my-sites/plans-grid/components/plan-type-selector/hooks/use-term-vise-max-dicounts.ts]: `NB:`

Would you mind explaining a bit more about the name "TermVise"? I'm not familiar with the concept

File: client/my-sites/plans-grid/components/plan-type-selector/hooks/use-interval-options.tsx [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+			key: 'monthly',
+			name: translate( 'Pay monthly' ),
+			discountText: '',
+			url: '',
+			termInMonths: 1,
+		},
+	};
+	const termViseDiscounts = useTermViseMaxDiscounts(
+		props.plans,
+		Object.keys( optionList ) as Array< UrlFriendlyTermType >,
```
Comment (jeyip) [inline client/my-sites/plans-grid/components/plan-type-selector/hooks/use-interval-options.tsx]: `NB:`

I think `SupportedUrlFriendlyTermType` is more accurate here?

```suggestion
		Object.keys( optionList ) as Array< SupportedUrlFriendlyTermType >,
```

File: client/my-sites/plans-grid/components/plan-type-selector/hooks/use-term-vise-max-dicounts.ts (hunk 3 above)
Comment (ddc22) [reply client/my-sites/plans-grid/components/plan-type-selector/hooks/use-term-vise-max-dicounts.ts]: It should be term wise. I am basically trying to calculate the maximum discount for each term.

Comment (ddc22) [reply client/my-sites/plans-grid/components/plan-type-selector/hooks/use-term-vise-max-dicounts.ts]: To elaborate...
Within a given term (Yearly, 2Yearly), I am trying to calculate the discount one can receive by switching to a higher tier plan.
So first what i do is convert the term price to monthly terms.

https://github.com/Automattic/wp-calypso/blob/5c292990ffeedf7054dfb6d9f6ff8dbb24ae0830/client/my-sites/plans-grid/components/plan-type-selector/hooks/use-term-vise-max-dicounts.ts#L87

Then I calculate the percentage difference in price compared to the monthly price.

https://github.com/Automattic/wp-calypso/blob/5c292990ffeedf7054dfb6d9f6ff8dbb24ae0830/client/my-sites/plans-grid/components/plan-type-selector/hooks/use-term-vise-max-dicounts.ts#L89

This is the discount that you can achieve by going with the higher term compared to the monthly cost.

Comment (jeyip) [reply client/my-sites/plans-grid/components/plan-type-selector/hooks/use-term-vise-max-dicounts.ts]: Let's rename the hook to `useMaxDiscountsForPlanTerms`. It seems like it better describes both the keys and values of the mapping returned by this hook. p1702325681579499/1702307941.915969-slack-CV2TX2PAN

File: client/my-sites/plans-grid/components/plan-type-selector/hooks/use-term-vise-max-dicounts.ts [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+			default:
+				return 1;
+		}
+	};
+
+	const termViseMaxDiscount: Record< UrlFriendlyTermType, number > = {} as Record<
+		UrlFriendlyTermType,
+		number
+	>;
+	termDefinitionsMapping.forEach( ( termMapping ) => {
```
Comment (jeyip) [inline client/my-sites/plans-grid/components/plan-type-selector/hooks/use-term-vise-max-dicounts.ts:72]: `NB:`

@jdc91 and I [agree that the density of operations and transformations needed here is surprising]( p1702325230911329/1702307941.915969-slack-CV2TX2PAN ), but I suppose when we give it more thought, calculating all possible discounts for all variations of each plan ( personal, premium, business, etc. ) from scratch could be a bit complex.

@jdc91 brought up a good point -- perhaps we could create an abstraction for the discount calculation for a single term. It's not something I'm convinced that we should immediately address since this is an experiment. @jdc91 and I _do_ believe, however, that if / when we do want to release this to a general audience, refactoring this block of code becomes a blocker.

File: client/my-sites/plans-grid/components/plan-type-selector/hooks/use-term-vise-max-dicounts.ts (hunk 3 above)
Comment (jeyip) [reply client/my-sites/plans-grid/components/plan-type-selector/hooks/use-term-vise-max-dicounts.ts]: > So first what i do is convert the term price to monthly terms.

OH! This is the step I was trying to wrap my head around. Do you mind adding an inline comment explaining this for now?

File: client/my-sites/plans-grid/components/plan-type-selector/hooks/use-interval-options.tsx [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+		props.usePricingMetaForGridPlans
+	);
+
+	const additionalPathProps = {
+		...( props.redirectTo ? { redirect_to: props.redirectTo } : {} ),
+		...( props.selectedPlan ? { plan: props.selectedPlan } : {} ),
```
Comment (jeyip) [inline client/my-sites/plans-grid/components/plan-type-selector/hooks/use-interval-options.tsx]: `NB:`

I might be missing something, but it looks like `selectedPlan` and `selectedFeature` are automatically destructured in `generatePath`. If that's correct, we don't need to include the two props in `additionalPathProps`

https://github.com/Automattic/wp-calypso/blob/21b708aa10610511be25f93ca76f48abd08050bc/client/my-sites/plans-grid/components/plan-type-selector/utils.ts#L15-L16

File: packages/calypso-products/src/get-plan-multiple-terms-variant-slugs.ts [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+	const plan = PLANS_LIST[ slug ];
+
+	if ( plan ) {
+		return Object.values( PLANS_LIST )
+			.filter( ( { type } ) => type === plan.type )
+			.filter( ( { term } ) => planTerms.includes( term ) )
+			.map( ( { getStoreSlug } ) => getStoreSlug() );
+	}
+
+	return [];
```
Comment (chriskmnds) [inline packages/calypso-products/src/get-plan-multiple-terms-variant-slugs.ts:24]: I feel null return is generally always better for empty (nothing found) results.

File: config/development.json [hunk 8]
```diff
@@ -141,6 +141,7 @@
 		"onboarding/import-from-wordpress": true,
 		"onboarding/import-light": false,
 		"onboarding/import-redirect-to-themes": true,
+		"onboarding/interval-dropdown": true,
```
Comment (chriskmnds) [inline config/development.json]: Why set it to `true` here? Do we all need to work with this experimental variant when developing other things (and have to turn it off) or just the folks working on this particular feature? I's pressume the latter.

File: client/my-sites/plans-grid/components/plan-type-selector/hooks/use-term-vise-max-dicounts.ts (hunk 3 above)
Comment (chriskmnds) [inline client/my-sites/plans-grid/components/plan-type-selector/hooks/use-term-vise-max-dicounts.ts]: I'vd found it hard to follow this function. Also throwing an Error feels overly severe to me. Not sure why that would ever happen or if worth to cripple the page and user experience when it does. At the very least I'd expect some documentation.

**Comment by jeyip** [comment]
## Testing
### Requirements

For each of the flows listed below:
- Can add a monthly plan to the cart
- Can add a yearly plan to the cart
- Can add a 2yearly plan to the cart
- Can add a 3yearly plan to the cart

Flows:
- [x] /start
- [x] /setup/newsletter

### Browsers
- [x] Chrome

### Notes
Because this feature is behind a feature flag, I'm fine moving forward with what we have and addressing any glaring issues in follow-ups 😎 

- Clicking the select dropdown in a certain location ( towards the center ) causes the page to refresh before selecting a dropdown item



- Enter keyboard button doesn't select a dropdown item



- Dropdown item hitbox seems small

File: client/my-sites/plans-grid/components/plan-type-selector/hooks/use-interval-options.tsx (hunk 6 above)
Comment (ddc22) [reply client/my-sites/plans-grid/components/plan-type-selector/hooks/use-interval-options.tsx]: Oh nice catch this was a copy paste from... 

https://github.com/Automattic/wp-calypso/blob/5c292990ffeedf7054dfb6d9f6ff8dbb24ae0830/client/my-sites/plans-grid/components/plan-type-selector/components/interval-type-toggle.tsx#L52-L56

File: client/my-sites/plans-grid/components/plan-type-selector/hooks/use-interval-options.tsx [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+		...( props.selectedFeature ? { feature: props.selectedFeature } : {} ),
+	};
+
+	const isDomainUpsellFlow = new URLSearchParams( window.location.search ).get( 'domain' );
+
+	const isDomainAndPlanPackageFlow = new URLSearchParams( window.location.search ).get(
+		'domainAndPlanPackage'
+	);
+
+	const isJetpackAppFlow = new URLSearchParams( window.location.search ).get( 'jetpackAppPlans' );
```
Comment (chriskmnds) [inline client/my-sites/plans-grid/components/plan-type-selector/hooks/use-interval-options.tsx]: Probably need to check for undefined `window` in this file. It will be undefined in case of SSR, and everything under plans-grid is subject to being so.

**Review by jeyip — APPROVED** [review]

File: client/my-sites/plans-grid/components/plan-type-selector/hooks/use-interval-options.tsx [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+
+	Object.keys( optionList ).forEach( ( key ) => {
+		optionList[ key as SupportedUrlFriendlyTermType ] = {
```
Comment (chriskmnds) [inline client/my-sites/plans-grid/components/plan-type-selector/hooks/use-interval-options.tsx]: `Object.fromEntries` seems to be the best practice nowadays for such transformations. So in this case, something like

```
Object.fromEntries(
  Object.keys( optionsList ).map( key => [ key, { ... } ] )
)
```

**Review by chriskmnds — COMMENTED** [review]
Some run-by comments. Mainly on setting the flag in development I was curious about. It seems like a lot of complexity at places though

File: config/development.json (hunk 8 above)
Comment (ddc22) [reply config/development.json]: Mainly set it true because the experimental variant should not be a blocker for anyone currently working on this area. And any of us folks who see this could reach out with actual feedback regarding the soon to be shipped feature. But happy to disable it if you think it could be a hindrance for others. 👍

Comment (chriskmnds) [reply config/development.json]: Ehm I don't know then. If this is part of experiment, I would rather we keep development environment closer to production.

Also, I think the flags API from calypso-config won't work outside of Calypso. So this is best passed through as a prop in plans-grid, as the goal is to use in other environments.

Comment (ddc22) [reply config/development.json]: Got it! Will do the changes!

File: client/my-sites/plans-grid/components/plan-type-selector/hooks/use-term-vise-max-dicounts.ts (hunk 3 above)
Comment (ddc22) [reply client/my-sites/plans-grid/components/plan-type-selector/hooks/use-term-vise-max-dicounts.ts]: I meant for this to be a dev error :) But you're correct the error is not required.

File: client/my-sites/plans-grid/components/plan-type-selector/hooks/use-interval-options.tsx (hunk 9 above)
Comment (ddc22) [reply client/my-sites/plans-grid/components/plan-type-selector/hooks/use-interval-options.tsx]: Good call will fix this.

File: client/my-sites/plans-grid/components/plan-type-selector/hooks/use-term-vise-max-dicounts.ts (hunk 3 above)
Comment (ddc22) [reply client/my-sites/plans-grid/components/plan-type-selector/hooks/use-term-vise-max-dicounts.ts]: > I'v found it hard to follow this function

I will add some documentation here so that it's not that complex. Also I am open to any suggestions to improve readability. The complexity here has me stumped as well.

File: client/my-sites/plans-grid/components/plan-type-selector/hooks/use-interval-options.tsx (hunk 10 above)
Comment (ddc22) [reply client/my-sites/plans-grid/components/plan-type-selector/hooks/use-interval-options.tsx]: But aren't we unnecessarily creating arrays here just to satisfy the from entries function? 🤔 
Curious to understand any advantage this has over the current approach? 👍

File: packages/calypso-products/src/get-plan-multiple-terms-variant-slugs.ts (hunk 7 above)
Comment (ddc22) [reply packages/calypso-products/src/get-plan-multiple-terms-variant-slugs.ts:24]: I actually returned an array here because it looks cleaner to chain this call where I was using it.
image

File: client/my-sites/plans-grid/components/plan-type-selector/hooks/use-interval-options.tsx (hunk 10 above)
Comment (chriskmnds) [reply client/my-sites/plans-grid/components/plan-type-selector/hooks/use-interval-options.tsx]: it's a tiny bit more performant than a `reduce` call with spreading, but certainly not a match performance-wise to mutating am object. I generally avoid mutations like fire, so up to taste

It's semantically more geared towards explaining the operation happening e.g. if you have an array of entries and you want to extract an object from it (then "fromEntries" captures that well). I think we'll be seeing a lot more of it going forward, so probably good to get into a habbit.

File: packages/calypso-products/src/get-plan-multiple-terms-variant-slugs.ts (hunk 7 above)
Comment (chriskmnds) [reply packages/calypso-products/src/get-plan-multiple-terms-variant-slugs.ts:24]: yeah that whole chain looks pretty awful 😆

Comment (ddc22) [reply packages/calypso-products/src/get-plan-multiple-terms-variant-slugs.ts:24]: Haha why is that? Happy to do any improvements to make it read better ! :)

Comment (ddc22) [reply packages/calypso-products/src/get-plan-multiple-terms-variant-slugs.ts:24]: Happy to pick this up later. Will merge for now.

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/10255400

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Thank you @jdc91 for including a screenshot in the description! This is really helpful for our translators.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/10255400).

File: client/my-sites/plans-grid/components/plan-type-selector/hooks/use-term-vise-max-dicounts.ts [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+import { UsePricingMetaForGridPlans } from 'calypso/my-sites/plans-grid/hooks/npm-ready/data-store/use-grid-plans';
+
+/**
+ * Calculate the maximum discount for each term for a given set of plans
+ * @param plans plans considered
+ * @param urlFriendlyTerms terms calculated for
+ * @param usePricingMetaForGridPlans price calculation hook
+ * @returns A map of term to maximum discount
+ */
+export default function useMaxDiscountsForPlanTerms(
```
Comment (chriskmnds) [inline client/my-sites/plans-grid/components/plan-type-selector/hooks/use-term-vise-max-dicounts.ts:20]: We changed the name here to `useMaxDiscountsForPlanTerms` (indeed better) but kept the filename `use-term-vise-max-dicounts` (it's also misspelled). We should rename the filename too cc @jeyip

Comment (jeyip) [reply client/my-sites/plans-grid/components/plan-type-selector/hooks/use-term-vise-max-dicounts.ts:20]: Good catch. I'll fix it now 👍

Comment (jeyip) [reply client/my-sites/plans-grid/components/plan-type-selector/hooks/use-term-vise-max-dicounts.ts:20]: https://github.com/Automattic/wp-calypso/pull/85505
